pkg/prometheus/promcfg.go

```diff
@@ -92,11 +92,28 @@ func NewConfigGenerator(logger log.Logger, p monitoringv1.PrometheusInterface, e
 		return nil, fmt.Errorf("failed to parse scrape classes: %w", err)
 	}
 
+	endpointSliceSupportedAndConfigured := false // Always assume false to preserve original prometheus-operator behaviour.
+
+	// Check if the user has explicitly set the service discovery role to use.
+	switch serviceDiscoveryRole := *cpf.ServiceDiscoveryRole; serviceDiscoveryRole {
+	case "EndpointSlice":
+		level.Info(logger).Log("msg", "using endpointslice as service discovery role")
+		endpointSliceSupportedAndConfigured = true
+	case "Endpoints":
+		level.Info(logger).Log("msg", "using endpoints as service discovery role")
+		endpointSliceSupportedAndConfigured = false
+	default:
+		level.Warn(logger).Log("msg",
+			"unknown service discovery role %q, defaulting to endpoints. Configure serviceDiscoveryRole to 'EndpointSlice' to use endpointslice as service discovery role.",
+			serviceDiscoveryRole)
+		endpointSliceSupportedAndConfigured = false
+	}
+
 	return &ConfigGenerator{
 		logger:                 logger,
 		version:                version,
 		prom:                   p,
-		endpointSliceSupported: endpointSliceSupported,
+		useEndpointSlice:       endpointSliceSupportedAndConfigured,
```

Comment:
What I'm imagining is that the argument `endpointSliceSupported` still exists, that tells if the cluster itself supports endpointSlice, and `endpointSliceConfigured` is what you're checking with the switch case above.

```suggestion
		useEndpointSlice:       endpointSliceSupported && endpointSliceConfigured,
```

Reply:
This is a wonderful comment that points out the missed logic in my implementation. I have fixed this in [16dc1f3](https://github.com/prometheus-operator/prometheus-operator/pull/6672/commits/16dc1f3d8a62d9076512e7e47993d7153177c4a9)